### PR TITLE
Accrue tests

### PR DESCRIPTION
**Pull request by MishaShWoof**
**Accrue Tests**

**`accrueAccount` test suite** — added a dedicated set of tests covering the `accrueAccount` function, verifying correct behavior across various collateralization states.

### CODE REVIEW

File: test/accrue-test.ts [hunk 1]
```diff
@@ -1,245 +1,558 @@
...
+
+  function calculateExpectedTrackingIndex(baseTrackingIndexBefore: BigNumber, trackingSpeed: BigNumber, timeElapsed: number, totalSupplyBase: BigNumber): BigNumber {
+    const indexAccrued = trackingSpeed.mul(timeElapsed).mul(baseTokenScale).div(totalSupplyBase);
+    return baseTrackingIndexBefore.add(indexAccrued);
+  }
+
+  function calculateExpectedUserTrackingAccrued(trackingIndexAfter: BigNumber, trackingIndexBefore: BigNumber, principal: BigNumber): BigNumber {
+    const indexDifference = trackingIndexAfter.sub(trackingIndexBefore);
+    return principal.mul(indexDifference).div(INDEX_SCALE);
+  }
```
Comment (VladyslavPerederWoof) [inline test/accrue-test.ts]: move to helpers. or calculate manually like we did here: https://github.com/woof-software/comet/blob/woof-software/supply-tests/test/interest-rate-test.ts

File: test/accrue-test.ts [hunk 2]
```diff
@@ -1,245 +1,558 @@
...
+
+  async function makeUtilization(_comet: CometHarnessInterfaceExtendedAssetList, utilization: number) {
+    if(utilization <= 0 || utilization >= 100) {
+      throw new Error('Utilization must be between 0 and 100');
+    }
```
Comment (VladyslavPerederWoof) [inline test/accrue-test.ts]: function create just for 2 edge cases and to make a specific utilization value on Comet. U can achieve that by doing that in the block for these cases. The function is too big and useless.

File: test/accrue-test.ts [hunk 3]
```diff
@@ -1,245 +1,558 @@
...
+  const INDEX_SCALE = exp(1, 15);
+  const baseTokenDecimals = 6;
+  const baseTokenScale = exp(1, baseTokenDecimals);
+  const seedAmount = exp(10_000, baseTokenDecimals);
+
+  const supplyAmount = exp(100, baseTokenDecimals);
+
+  const oneDay = 24 * 60 * 60;
+  const oneMonth = 30 * oneDay;
+
```
Comment (VladyslavPerederWoof) [inline test/accrue-test.ts]: Move to helper. If I'm not mistake we'll use these values in other tests.

File: test/accrue-test.ts [hunk 4]
```diff
@@ -1,245 +1,558 @@
...
-  it('accrues correctly with time elapsed and less than min rewards', async () => {
-    await ethers.provider.send('hardhat_reset', []); // ensure clean start...
+    it('supply and borrow indexes should be 1 in initial state', async function () {
+      const { baseSupplyIndex, baseBorrowIndex } = await comet.totalsBasic();
+      // initial index should be index
+      const indexScale = await comet.baseIndexScale();
+      expect(baseSupplyIndex).to.be.equal(INDEX_SCALE);
+      expect(baseSupplyIndex).to.be.equal(indexScale);
+      expect(baseBorrowIndex).to.be.equal(indexScale);
+    });
```
Comment (VladyslavPerederWoof) [inline test/accrue-test.ts]: Split to separate tests. supply index and borrow

File: test/accrue-test.ts [hunk 5]
```diff
@@ -1,245 +1,558 @@
...
-    await ethers.provider.send('evm_setAutomine', [false]);
-    const _a1 = await comet.accrue();
-    await ethers.provider.send('evm_mine', [start + 1000]);
-    await ethers.provider.send('evm_setAutomine', [true]);
+      const expectedSupplyIndexAccrued = BigNumber.from(INDEX_SCALE).mul(supplyPerSecond).mul(timeElapsed).div(factorScale);
+      const expectedSupplyIndex = BigNumber.from(INDEX_SCALE).add(expectedSupplyIndexAccrued);
 
-    const t2 = await comet.totalsBasic();
+      const expectedBorrowIndexAccrued = BigNumber.from(INDEX_SCALE).mul(borrowPerSecond).mul(timeElapsed).div(factorScale);
+      const expectedBorrowIndex = BigNumber.from(INDEX_SCALE).add(expectedBorrowIndexAccrued);
```
Comment (VladyslavPerederWoof) [inline test/accrue-test.ts:152]: Split to separate tests. supply index and borrow

File: test/accrue-test.ts [hunk 6]
```diff
@@ -1,245 +1,558 @@
...
-    expect(timeElapsed).to.be.equal(1000);
+      expect(baseSupplyIndex).to.equal(expectedSupplyIndex);
+      expect(baseBorrowIndex).to.equal(expectedBorrowIndex);
+    });
+
+    it('total supply and borrow should remain 0 after accrue', async function () {
+      const { totalSupplyBase, totalBorrowBase } = await comet.totalsBasic();
+      expect(totalSupplyBase).to.equal(0);
+      expect(totalBorrowBase).to.equal(0);
+    });
```
Comment (VladyslavPerederWoof) [inline test/accrue-test.ts]: Split to separate tests (supply and borrow)

File: test/accrue-test.ts [hunk 7]
```diff
@@ -1,245 +1,558 @@
...
-    expect(t2.baseSupplyIndex).to.be.equal(projectBaseIndex(t1.baseSupplyIndex, supplyRate, timeElapsed));
-    expect(t2.baseBorrowIndex).to.be.equal(projectBaseIndex(t1.baseBorrowIndex, borrowRate, timeElapsed));
-    expect(t2.trackingSupplyIndex).to.be.equal(t1.trackingSupplyIndex);
-    expect(t2.trackingBorrowIndex).to.be.equal(t1.trackingBorrowIndex);
+    it('user position is not affected by accrue with no activity', async function () {
+      const userBasicAfter = await comet.userBasic(alice.address);
+      expect(userBasicAfter).to.deep.equal(userBasicBefore);
+    });
+
+    describe('seeding reserves', function () {
```
Comment (VladyslavPerederWoof) [inline test/accrue-test.ts:172]: Need more informative description.
seeding reserves impact on ...., separate describe block seeding reserves do not impact on ....

also after allocation make spending time and mine block to make sure that time impact for certain things.

File: test/accrue-test.ts [hunk 8]
```diff
@@ -1,245 +1,558 @@
...
+        expect(totalsAfter.trackingBorrowIndex).to.equal(totalsBefore.trackingBorrowIndex);
+        expect(totalsAfter.totalSupplyBase).to.equal(totalsBefore.totalSupplyBase);
+        expect(totalsAfter.totalBorrowBase).to.equal(totalsBefore.totalBorrowBase);
+
+        const expectedSupplyIndex = calculateExpectedIndex(totalsBefore.baseSupplyIndex, supplyRatePerSecond, 2);
+        const expectedBorrowIndex = calculateExpectedIndex(totalsBefore.baseBorrowIndex, borrowRatePerSecond, 2);
+
+        expect(totalsAfter.baseSupplyIndex).to.equal(expectedSupplyIndex);
+        expect(totalsAfter.baseBorrowIndex).to.equal(expectedBorrowIndex);
+      });
```
Comment (VladyslavPerederWoof) [inline test/accrue-test.ts]: This test shows that tracking indexes and total supply and borrow not changed and in the same test describe 0 info that base and borrow index changed. split tests.
Also totalsAfter taken here, after splitting it's better to move right after allocating in before block.

File: test/accrue-test.ts [hunk 9]
```diff
@@ -1,245 +1,558 @@
...
+      const borrowRatePerSecond = BigNumber.from(config.borrowInterestRateBase).div(365 * oneDay);
+
+      expect(totalsAfter.baseSupplyIndex).to.equal(calculateExpectedIndex(totalsBefore.baseSupplyIndex, supplyRatePerSecond, 3));
+      expect(totalsAfter.baseBorrowIndex).to.equal(calculateExpectedIndex(totalsBefore.baseBorrowIndex, borrowRatePerSecond, 3));
+
+      expect(totalsAfter.trackingSupplyIndex).to.equal(totalsBefore.trackingSupplyIndex);
+      expect(totalsAfter.trackingBorrowIndex).to.equal(totalsBefore.trackingBorrowIndex);
+      expect(totalsAfter.totalSupplyBase).to.be.closeTo(supplyAmount, 1);
+      expect(totalsAfter.totalBorrowBase).to.equal(totalsBefore.totalBorrowBase);
+    });
```
Comment (VladyslavPerederWoof) [inline test/accrue-test.ts]: same here. split. some values are changed and some not. Split tests for more informative

File: test/accrue-test.ts [hunk 10]
```diff
@@ -1,245 +1,558 @@
...
+      const supplyRate = await comet.getSupplyRate(0);
+      const borrowRatePerSecond = BigNumber.from(config.borrowInterestRateBase).div(365 * oneDay);
+
+      expect(totalsAfter.baseSupplyIndex).to.be.closeTo(calculateExpectedIndex(totalsBefore.baseSupplyIndex, supplyRate, oneMonth), supplyRate.div(100));
+      expect(totalsAfter.baseBorrowIndex).to.be.closeTo(calculateExpectedIndex(totalsBefore.baseBorrowIndex, borrowRatePerSecond, oneMonth), borrowRatePerSecond.div(100));
+
+      expect(totalsAfter.trackingSupplyIndex).to.equal(calculateExpectedTrackingIndex(BigNumber.from(0), BigNumber.from(config.baseTrackingSupplySpeed), oneMonth + 1, totalsAfter.totalSupplyBase));
+      expect(totalsAfter.totalSupplyBase).to.be.closeTo(supplyAmount, 1);
+      expect(totalsAfter.totalBorrowBase).to.equal(totalsBefore.totalBorrowBase);
+    });
```
Comment (VladyslavPerederWoof) [inline test/accrue-test.ts]: same here. split. some values are changed and some not. Split tests for more informative

File: test/accrue-test.ts [hunk 11]
```diff
@@ -1,245 +1,558 @@
...
-
-    expect(t2.baseSupplyIndex).to.be.equal(projectBaseIndex(t1.baseSupplyIndex, supplyRate, timeElapsed));
-    expect(t2.baseBorrowIndex).to.be.equal(projectBaseIndex(t1.baseBorrowIndex, borrowRate, timeElapsed));
-    expect(t2.trackingSupplyIndex).to.be.equal(t1.trackingSupplyIndex);
-    expect(t2.trackingBorrowIndex).to.be.equal(t1.trackingBorrowIndex);
+  describe('borrowing position', function () {
+    const borrowAmount = exp(100, baseTokenDecimals);
+
+    let totalsBefore: TotalsBasicStructOutput;
+    let userBasicBefore: [BigNumber, BigNumber, BigNumber, number, number] & { principal: BigNumber, baseTrackingIndex: BigNumber, baseTrackingAccrued: BigNumber, assetsIn: number, _reserved: number };
```
Comment (VladyslavPerederWoof) [inline test/accrue-test.ts]: move to helpers for future re-using

File: test/accrue-test.ts [hunk 12]
```diff
@@ -1,245 +1,558 @@
...
+    before(async function () {
+      // Bob supplies 1 WETH as collateral
+      await collaterals.WETH.connect(bob).allocateTo(bob.address, exp(1, 18));
+      await collaterals.WETH.connect(bob).approve(comet.address, exp(1, 18));
+      await comet.connect(bob).supply(collaterals.WETH.address, exp(1, 18));
+    });
+
+    it('should allow borrowing', async function () {
+      expect(await comet.connect(bob).withdraw(baseToken.address, borrowAmount)).to.not.be.reverted;
+    });
```
Comment (VladyslavPerederWoof) [inline test/accrue-test.ts]: move to before

File: test/accrue-test.ts [hunk 13]
```diff
@@ -1,245 +1,558 @@
...
+      expect(utilization).to.be.gt(0);
+    });
+
+    it('should allow accruing with utilization > 0', async function () {
+      totalsBefore = await comet.totalsBasic();
+      userBasicBefore = await comet.userBasic(bob.address);
+
+      const utilization = await comet.getUtilization();
+      supplyRateBefore = await comet.getSupplyRate(utilization);
+      borrowRateBefore = await comet.getBorrowRate(utilization);
```
Comment (VladyslavPerederWoof) [inline test/accrue-test.ts:365]: move to before after withdraw. tests will be more clean

File: test/accrue-test.ts [hunk 14]
```diff
@@ -1,245 +1,558 @@
...
+
+      // Fast forward a month
+      await ethers.provider.send('evm_increaseTime', [oneMonth]);
+      await ethers.provider.send('evm_mine', []);
```
Comment (VladyslavPerederWoof) [inline test/accrue-test.ts]: make separate tests above this test. (skip 1 month to accrue interest)

File: test/accrue-test.ts [hunk 15]
```diff
@@ -1,245 +1,558 @@
...
+      const timeElapsed = totalsAfter.lastAccrualTime - totalsBefore.lastAccrualTime;
+      expect(totalsAfter.baseSupplyIndex).to.be.equal(calculateExpectedIndex(totalsBefore.baseSupplyIndex, supplyRateBefore, timeElapsed));
+      expect(totalsAfter.baseBorrowIndex).to.be.equal(calculateExpectedIndex(totalsBefore.baseBorrowIndex, borrowRateBefore, timeElapsed));
+
+      expect(totalsAfter.trackingSupplyIndex).to.equal(calculateExpectedTrackingIndex(totalsBefore.trackingSupplyIndex, BigNumber.from(config.baseTrackingSupplySpeed), timeElapsed, totalsBefore.totalSupplyBase));
+      expect(totalsAfter.trackingBorrowIndex).to.equal(calculateExpectedTrackingIndex(totalsBefore.trackingBorrowIndex, BigNumber.from(config.baseTrackingBorrowSpeed), timeElapsed, totalsBefore.totalBorrowBase));
+
+      expect(totalsAfter.totalSupplyBase).to.equal(totalsBefore.totalSupplyBase);
+      expect(totalsAfter.totalBorrowBase).to.equal(totalsBefore.totalBorrowBase);    
+    });
```
Comment (VladyslavPerederWoof) [inline test/accrue-test.ts]: same here. split. some values are changed and some not. Split tests for more informative

File: test/accrue-test.ts [hunk 16]
```diff
@@ -1,245 +1,558 @@
...
+      const totalsAfter = await comet.totalsBasic();
+
+      const expectedBorrowTrackingIndex = calculateExpectedTrackingIndex(userBasicBefore.baseTrackingIndex, BigNumber.from(config.baseTrackingBorrowSpeed), oneMonth + 1, totalsAfter.totalBorrowBase);
+      const expectedBorrowTrackingAccrued = calculateExpectedUserTrackingAccrued(totalsAfter.trackingBorrowIndex, userBasicBefore.baseTrackingIndex, userBasicBefore.principal.mul(-1));
+
+      expect(userBasicAfter.principal).to.be.equal(userBasicBefore.principal);
+      expect(userBasicAfter.baseTrackingIndex).to.be.closeTo(expectedBorrowTrackingIndex, 1);
+      expect(userBasicAfter.baseTrackingAccrued).to.equal(expectedBorrowTrackingAccrued);
+      expect(userBasicAfter.assetsIn).to.equal(userBasicBefore.assetsIn);
+    });
```
Comment (VladyslavPerederWoof) [inline test/accrue-test.ts]: same here. split. some values are changed and some not. Split tests for more informative

File: test/accrue-test.ts [hunk 17]
```diff
@@ -1,245 +1,558 @@
...
 
-    const t0 = await comet.totalsBasic();
-    const t1 = Object.assign({}, t0, {
-      baseSupplyIndex: 2n ** 64n - 1n,
-      totalSupplyBase: 14000,
-      totalBorrowBase: 13000, // needs to have positive utilization for supply rate to be > 0
+    it('borrow balance should be 0 after repay', async function () {
+      const borrowBalance = await comet.borrowBalanceOf(bob.address);
+      expect(borrowBalance).to.equal(0);
```
Comment (VladyslavPerederWoof) [inline test/accrue-test.ts]: can be in single line

File: test/accrue-test.ts [hunk 18]
```diff
@@ -1,245 +1,558 @@
...
-    const _s0 = await wait(comet.setTotalsBasic(t1));
-    await fastForward(2);
-    await expect(wait(comet.accrue())).to.be.revertedWith('code 0x11 (Arithmetic operation underflowed or overflowed outside of an unchecked block)');
 
-    const t2 = Object.assign({}, t0, {
-      baseBorrowIndex: 2n ** 64n - 1n,
+    it('utilization should be 0 after repay', async function () {
+      const utilization = await comet.getUtilization();
+      expect(utilization).to.equal(0);
+    });
```
Comment (VladyslavPerederWoof) [inline test/accrue-test.ts:460]: single line

File: test/accrue-test.ts [hunk 19]
```diff
@@ -1,245 +1,558 @@
...
+
+    it('user position should change to supply after repay', async function () {
+      const userBasic = await comet.userBasic(bob.address);
+      expect(userBasic.principal).to.be.gt(0);
+    });
```
Comment (VladyslavPerederWoof) [inline test/accrue-test.ts:465]: single line

File: test/accrue-test.ts [hunk 20]
```diff
@@ -1,245 +1,558 @@
...
+
+      expect(totalsAfter.baseSupplyIndex).to.be.equal(calculateExpectedIndex(totalsBefore.baseSupplyIndex, await comet.getSupplyRate(0), 1));
+      expect(totalsAfter.baseBorrowIndex).to.be.equal(calculateExpectedIndex(totalsBefore.baseBorrowIndex, await comet.getBorrowRate(0), 1));
+
+      expect(totalsAfter.trackingSupplyIndex).to.equal(calculateExpectedTrackingIndex(totalsBefore.trackingSupplyIndex, BigNumber.from(config.baseTrackingSupplySpeed), 1, totalsBefore.totalSupplyBase));
+      expect(totalsAfter.trackingBorrowIndex).to.equal(totalsBefore.trackingBorrowIndex);
+
+      expect(userBasicAfter.principal).to.equal(userBasicBefore.principal);
+      expect(userBasicAfter.baseTrackingIndex).to.be.closeTo(userBasicBefore.baseTrackingIndex, 2e8);
+      expect(userBasicAfter.assetsIn).to.equal(userBasicBefore.assetsIn);
```
Comment (VladyslavPerederWoof) [inline test/accrue-test.ts]: same here. split. some values are changed and some not. Split tests for more informative

File: test/accrue-test.ts [hunk 21]
```diff
@@ -1,245 +1,558 @@
...
+      expect(totalsAfter.baseBorrowIndex).to.be.equal(calculateExpectedIndex(totalsBefore.baseBorrowIndex, await comet.getBorrowRate(0), 1));
+
+      expect(totalsAfter.trackingSupplyIndex).to.equal(calculateExpectedTrackingIndex(totalsBefore.trackingSupplyIndex, BigNumber.from(config.baseTrackingSupplySpeed), 1, totalsBefore.totalSupplyBase));
+      expect(totalsAfter.trackingBorrowIndex).to.equal(totalsBefore.trackingBorrowIndex);
+
+      expect(userBasicAfter.principal).to.equal(_userBasicBefore.principal);
+      expect(userBasicAfter.baseTrackingIndex).to.equal(totalsAfter.trackingSupplyIndex);
+      expect(userBasicAfter.baseTrackingAccrued).to.equal(_userBasicBefore.baseTrackingAccrued);
+      expect(userBasicAfter.assetsIn).to.equal(_userBasicBefore.assetsIn);
+    });
```
Comment (VladyslavPerederWoof) [inline test/accrue-test.ts:721]: same here. split. some values are changed and some not. Split tests for more informative

File: test/accrue-test.ts [hunk 22]
```diff
@@ -1,245 +1,558 @@
...
   });
-});
 
-describe('accrueAccount', function () {
-  it('has no effect when called on an address with no protocol activity', async () => {
-    const { comet, users: [unusedAccount] } = await makeProtocol();
+  describe('edge cases', function () {
+    let snapshot: string;
+    beforeEach(async function () {
+      snapshot = await ethers.provider.send('evm_snapshot', []);
```
Comment (VladyslavPerederWoof) [inline test/accrue-test.ts]: use snapshot.ts

File: test/accrue-test.ts [hunk 23]
```diff
@@ -1,245 +1,558 @@
...
+    });
 
-    const userBasic0 = await comet.userBasic(unusedAccount.address);
-    await comet.accrueAccount(unusedAccount.address);
-    const userBasic1 = await comet.userBasic(unusedAccount.address);
+    it('should revert when skipping too much time', async function () {
+      await ethers.provider.send('evm_increaseTime', [Number(exp(1, 18))]);
```
Comment (VladyslavPerederWoof) [inline test/accrue-test.ts]: use uint40.max - current time stamp + 1 to skip exact time

File: test/accrue-test.ts [hunk 24]
```diff
@@ -1,245 +1,558 @@
...
+              liquidateCF: exp(0.95, 18),
+              liquidationFactor: exp(0.95, 18),
+            },
+          }),
+        },
+      );
+      const comet = protocol.cometWithExtendedAssetList;
+      await makeUtilization(comet, 50);
+
+      await ethers.provider.send('evm_increaseTime', [10 * 365 * oneDay]);
```
Comment (VladyslavPerederWoof) [inline test/accrue-test.ts]: I guess we can also make a calculation to get the exact value of how much time is needed to skip here

File: test/accrue-test.ts [hunk 25]
```diff
@@ -1,245 +1,558 @@
...
```
Comment (VladyslavPerederWoof) [inline test/accrue-test.ts]: I guess we can also make a calculation to get the exact value of how much time is needed to skip here

**Comment by VladyslavPerederWoof** [comment]
1. accrueAccount idempotency within the same block
Calling accrueAccount twice in the same block should be a pure no-op (timeElapsed = 0). Not tested.  

2. accrueAccount on one user after global state was advanced by another user's action                                                                                                 
Pattern: bob's supply/withdraw triggers accrueInternal (global), updating lastAccrualTime. Then accrueAccount(alice) is called. Alice's baseTrackingAccrued should update based on the delta between totals.trackingSupplyIndex and alice.baseTrackingIndex. No test isolates this sequence explicitly.

3. balanceOf / borrowBalanceOf increasing over time                                                                                                                                   
The test verifies userBasic.principal is unchanged after accrue (correct — principal is fixed). But there is no test confirming that balanceOf(alice) (which calls presentValueSupply(baseSupplyIndex, principal)) actually returns a larger value after the index grows, i.e., that supplier earned interest is reflected in their balance.

4. lastAccrualTime after accrueAccount vs. state-changing operations                                                                                                                 
accrueInternal updates lastAccrualTime. The test checks lastAccrualTime after supply (line 264-268) but never after a standalone accrueAccount. There's no explicit assertion that lastAccrualTime = block.timestamp after accrueAccount.

**Comment by VladyslavPerederWoof** [comment]
Closed after merge dev into service patch
